src/stdhasher.cr

```diff
@@ -0,0 +1,257 @@
+  SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
+  end
+
+  protected def initialize(a, b)
+    @h = Pointer(Impl).malloc(1, Impl.new(a, b))
+  end
+
+  def self.build
+    h = Impl.new(@@seed[0], @@seed[1])
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def self.hashit(v)
+    build do |s|
+      s << v
+    end
+  end
+
+  protected def self.build(a, b)
+    h = Impl.new(a, b)
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def clone_build
+    self.class.build(@h.value.a, @h.value.b) do |s|
+      yield s
+    end
+  end
+
+  def clone
+    self.class.new(@h.value.a, @h.value.b)
+  end
+
+  def <<(v : Nil)
+    @h.value.permute_nil(@@seed[2])
+    self
+  end
+
+  # mix raw value without number normalizing
+  def raw(v : Int8 | Int16 | Int32 | UInt8 | UInt16 | UInt32)
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  # mix raw value without number normalizing
+  def raw(v : Int64 | UInt64)
+    @h.value.permute((v >> 32).to_u32, @@seed[2])
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  def <<(v : Int8 | Int16 | UInt8 | UInt16)
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  def <<(b : Bytes)
+    @h.value.permute(b, @@seed[2])
+    self
+  end
+
+  def <<(v)
+    v.hash(self)
+    self
+  end
+
+  def remix(v : Value)
+    self << Raw32.new(v)
+  end
+
+  def digest
+    @h.value.digest(@@seed[3])
+  end
+
+  private struct Impl
+    getter a : UInt32 = 0_u32
+    getter b : UInt32 = 0_u32
+
+    def initialize(@a : UInt32, @b : UInt32)
+    end
+
+    def permute(v : UInt32, s : UInt32)
+      permute(v, s, pointerof(@a), pointerof(@b))
+    end
+
+    def permute_nil(s : UInt32)
+      @a += s | 1
+      # LFSR
+      mx = (@b.to_i32 >> 31).to_u32 & 0xa8888eef_u32
+      @b = (@b << 1) ^ mx
+    end
+
+    def digest(seed)
+      a, b = @a, @b
+      b += seed
+      a ^= a >> 15
+      b ^= b >> 16
+      a *= 0xb8b34b2d_u32
+      b *= 0x52c6a2d9_u32
+      a ^= a >> 17
+      b ^= b >> 16
+      b + a
+    end
+
+    def permute(buf : Bytes, s : UInt32)
+      bsz = buf.size
+      v = bsz.to_u32 << 24
+      u = buf.to_unsafe
+      a, b = @a, @b
+      i = bsz.unsafe_div 4
+      while i > 0
+        permute(u.as(Pointer(UInt32)).value, s, pointerof(a), pointerof(b))
+        u += 4
+        i -= 1
+      end
+      r = (bsz & 3).to_u32
+      if r != 0
+        v |= u[0].to_u32 | (u[r/2].to_u32 << 8) | (u[r - 1].to_u32 << 16)
+      end
+      permute(v, s, pointerof(a), pointerof(b))
+      @a, @b = a, b
+      self
+    end
+
+    private def permute(v : UInt32, s : UInt32, aa : Pointer(UInt32), bb : Pointer(UInt32))
+      a, b = aa.value, bb.value
+      v ^= s
+      v *= 0xb8b34b2d_u32
+      a += v
+      a = rotl(a, 13)
+      b ^= a + s
+      b *= 9
+      aa.value, bb.value = a, b
```

Comment:
I use local variables instead of instance variables to reduce redundant stores to instance variables. I looked to assembly and looks like Crystal/LLVM is not aggressive enought in removing this stores. Especially it is visible in a loop of Bytes permutation.

I use pointers, cause LLVM does bad when to 32bit variables are packed into struct/tuple - it combines them into 64bit argument/result, and function inlining is not able to unsplit them. It is better said, LLVM lacks 'local var struct decomposition' (i don't remember exact name) ie when struct is function local (maybe after inlining), its fields become independent variables. GCC has such optimisation.

Macro is a great advise! Thanks! I'll try it.